### PR TITLE
EAR-677-validate-or-on-checkboxes-in-logic

### PR DESCRIPTION
**Pull request by MattElcock**
### What is the context of this PR?

This pull request improves upon existing validation for mutually exclusive checkboxes in routing. The backend has been streamlined and bugs removed, whilst the front end has been enhanced by the addition of red borders around included fields to help the user notice what needs to be changed.

### How to review 

Using the ACs on the Jira ticket, ensure that validation states can be correctly reached and are are not seen in other circumstances where an error being shown is incorrect.

* Import the **Capability examples** questionnaire from pre-prod Author into your local environment and:
    * ensure it can be opened in Author;
    * then, ensure it can be viewed in Runner by pressing the **view survey** button
* Does this require a migration? We need one if existing JSON schema properties change

### What to do after everything is green

1. * [ ] Bring this branch up-to-date with Origin/Master
2. * [ ] If there are a lot of commits or some are not helpful to read, squash them down
3. * [ ] Click the **Merge** button on this pull request
4. * [ ] Does this change mean the **Capability examples** questionnaire should be updated?
5. * [ ] Move the Jira ticket for this task into the next stage of the process

### CODE REVIEW

File: eq-author-api/src/validation/createValidationError.js [hunk 1]
```diff
@@ -113,6 +113,10 @@ module.exports = (dataPath, field, errorCode, questionnaire) => {
           rule = rules[dataPath[index + 2]];
           validationErr.routingRuleId = rule.id;
           validationErr.type = "routing";
+
+          if (rule.expressionGroup) {
+            validationErr.expressionGroupId = rule.expressionGroup.id;
+          }
```
Comment (tjamesmac) [inline eq-author-api/src/validation/createValidationError.js:119]: Does this fix that caching issue we were having?

File: eq-author/src/App/page/Logic/BinaryExpressionEditor/MultipleChoiceAnswerOptionsSelector.js [hunk 2]
```diff
@@ -201,23 +205,26 @@ class MultipleChoiceAnswerOptionsSelector extends React.Component {
...
           data-test="options-selector"
-          hasError={hasError}
+          hasError={errors.length}
         >
           <Label>Match</Label>
           <ConditionSelect
             onChange={this.handleConditionChange}
             defaultValue={expression.condition}
+            hasError={
+              errors.filter(({ field }) => field === "condition").length
```
Comment (tjamesmac) [inline eq-author/src/App/page/Logic/BinaryExpressionEditor/MultipleChoiceAnswerOptionsSelector.js:226]: This line is causing the coverage to drop

File: eq-author/src/App/page/Logic/Routing/RoutingPage/RoutingEditor/RuleEditor/index.js [hunk 3]
```diff
@@ -116,12 +116,27 @@ export class UnwrappedRuleEditor extends React.Component {
...
+      }
+      const expressionGroupOperatorErrors = validationErrorInfo.errors.filter(
+        ({ field }) => field === "groupOperator"
+      );
+
+      if (expressionGroupOperatorErrors.length) {
+        return true;
+      }
+
+      return false;
```
Comment (tjamesmac) [inline eq-author/src/App/page/Logic/Routing/RoutingPage/RoutingEditor/RuleEditor/index.js:137]: This is also causing the coverage to drop

File: eq-author-api/schema/resolvers/logic/routing2/expressionGroup2/index.js [hunk 4]
```diff
@@ -7,6 +7,28 @@ const { getPages } = require("../../../utils");
...
+    );
+
+    if (!expressionGroupErrors) {
+      const noErrors = {
+        id,
+        errors: [],
+        totalCount: 0,
+      };
+      return noErrors;
+    }
```
Comment (tjamesmac) [inline eq-author-api/schema/resolvers/logic/routing2/expressionGroup2/index.js:22]: This is also dropping the coverage

File: eq-author/src/constants/validationMessages.js [hunk 5]
```diff
@@ -95,9 +95,15 @@ export const rightSideErrors = {
...
-    message: "All of not allowed in expression with checkbox OR option",
+  ERR_RIGHTSIDE_MIXING_OR_STND_OPTIONS_IN_AND_RULE: {
+    errorCode: "ERR_RIGHTSIDE_MIXING_OR_STND_OPTIONS_IN_AND_RULE",
+    message:
+      "You can't match 'all of' the standard options with the 'or' checkbox option. Change the 'All of' to 'Any of' or remove all of the standard options or the 'or' checkbox option",
+  },
+  ERR_GROUP_MIXING_EXPRESSIONS_WITH_OR_STND_OPTIONS_IN_AND: {
+    errorCode: "ERR_GROUP_MIXING_EXPRESSIONS_WITH_OR_STND_OPTIONS_IN_AND",
+    message:
+      "You can't match 'all of' the standard options with the 'or' checkbox option. Change the 'All of' to 'Any of' or remove a condition",
```
Comment (tjamesmac) [inline eq-author/src/constants/validationMessages.js:106]: Do the first 'all of' need to be capitalised like the second ones?

**Review by tjamesmac — COMMENTED** [review]
Is the error message meant to be present in this situation? Neither of them are `or`
Screenshot 2020-09-24 at 09 45 14

This screenshot is trying to emulate AC 4 but it doesn't get an error message. When I match it to the design by adding another rule it adds an error though.
Screenshot 2020-09-24 at 09 48 29


Other than that, that's some super nice work!

**Comment by MattElcock** [comment]
> Is the error message meant to be present in this situation? Neither of them are `or`
> Screenshot 2020-09-24 at 09 45 14
> 
> This screenshot is trying to emulate AC 4 but it doesn't get an error message. When I match it to the design by adding another rule it adds an error though.
> Screenshot 2020-09-24 at 09 48 29
> 
> Other than that, that's some super nice work!

The screenshots aren't exactly the same. In the AC one they're both 'All of' where yours is 'all of' and 'any of'.

File: eq-author-api/src/validation/createValidationError.js (hunk 1 above)
Comment (MattElcock) [reply eq-author-api/src/validation/createValidationError.js:119]: No, that's just to help make it easier to catch expression group errors, 'cause some of the AC for this ticket is taking multiple rules in a group into account.

File: eq-author/src/constants/validationMessages.js (hunk 5 above)
Comment (MattElcock) [reply eq-author/src/constants/validationMessages.js:106]: Yes, it does! Sorry.

**Comment by MattElcock** [comment]
> Is the error message meant to be present in this situation? Neither of them are `or`
> Screenshot 2020-09-24 at 09 45 14
> 
> This screenshot is trying to emulate AC 4 but it doesn't get an error message. When I match it to the design by adding another rule it adds an error though.
> Screenshot 2020-09-24 at 09 48 29
> 
> Other than that, that's some super nice work!

Oop, okay! I'll take a look at that today.

File: eq-author-api/schema/resolvers/logic/routing2/expressionGroup2/index.js (hunk 4 above)
Comment (MattElcock) [reply eq-author-api/schema/resolvers/logic/routing2/expressionGroup2/index.js:22]: Weird, I've put a test in for that. I'll take a look.

Comment (MattElcock) [reply eq-author-api/schema/resolvers/logic/routing2/expressionGroup2/index.js:22]: Had a look and I'm not sure why this isn't being covered. The test is in `schema/tests/routing.test.js`

**Comment by MattElcock** [comment]
> > Is the error message meant to be present in this situation? Neither of them are `or`
> > Screenshot 2020-09-24 at 09 45 14
> > This screenshot is trying to emulate AC 4 but it doesn't get an error message. When I match it to the design by adding another rule it adds an error though.
> > Screenshot 2020-09-24 at 09 48 29
> > Other than that, that's some super nice work!
> 
> Oop, okay! I'll take a look at that today.

Fixed this.

File: eq-author/src/App/page/Logic/Routing/RoutingPage/RoutingEditor/RuleEditor/index.js (hunk 3 above)
Comment (MattElcock) [reply eq-author/src/App/page/Logic/Routing/RoutingPage/RoutingEditor/RuleEditor/index.js:137]: I've put in a test for this.

**Review by tjamesmac — DISMISSED** [review]
Beautiful

File: eq-author/src/App/page/Logic/BinaryExpressionEditor/MultipleChoiceAnswerOptionsSelector.js [hunk 6]
```diff
@@ -157,8 +156,13 @@ class MultipleChoiceAnswerOptionsSelector extends React.Component {
   handleError = () => {
     const { expression } = this.props;
 
-    const { errorCode } = find(expression.validationErrorInfo.errors, error =>
-      error.errorCode.includes("ERR_RIGHTSIDE")
+    const { errorCode } = find(
+      expression.validationErrorInfo.errors,
+      error =>
+        error.errorCode.includes("ERR_RIGHTSIDE") ||
+        error.errorCode.includes(
```
Comment (zimboflyman) [inline eq-author/src/App/page/Logic/BinaryExpressionEditor/MultipleChoiceAnswerOptionsSelector.js:163]: There's a coverage drop on this line

**Review by zimboflyman — COMMENTED** [review]
All looks really really good! - just a bit of test coverage could possibly be added?

**Review by zimboflyman — APPROVED** [review]
All looks really really good! - just a bit of test coverage could possibly be added?

**Comment by tjamesmac** [comment]
Not quite sure how the coverage dropped after merging master. Gonna give it a tick though, nice one!

**Review by tjamesmac — APPROVED** [review]

**Review by zimboflyman — APPROVED** [review]
nice!

**Review by jasehumphr — APPROVED** [review]
Sweet

**Pull request merged**